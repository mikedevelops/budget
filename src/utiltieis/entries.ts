import { IFigureDictionary, ICurrencyFigure } from '../interfaces/Figure';

/**
 * Merge Currecy figures
 * @param figureOld
 * @param figureNew
 */
export function mergeCurrencyFigure (
    figureOld: ICurrencyFigure,
    figureNew: ICurrencyFigure
): ICurrencyFigure {
    figureOld.value += figureNew.value;
    return figureOld;
}

export function getCumulativeEntries (
    entries: IFigureDictionary[]
): IFigureDictionary {
    return entries.reduce((result: IFigureDictionary, entry: IFigureDictionary) => {
        Object.keys(entry).forEach((figureKey: string) => {
            const unique = !result.hasOwnProperty(figureKey);
            const figure = entry[figureKey];

            // If the entry is unique to the cumulative object
            // we will just assignte new property and return
            if (unique) {
                result[figureKey] = figure;
                return result;
            }

            // Handle currency figures
            if (figure.type === 'currency') {
                result[figureKey] = mergeCurrencyFigure(result[figureKey], figure);
                return result;
            }
        });

        return result;
    }, {});
}
