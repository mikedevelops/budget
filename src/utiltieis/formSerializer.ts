export interface ISerializedForm {
    [index: string]: any;
}

export function serialize (
    data: FormData
): ISerializedForm {
    const formData: ISerializedForm = {};

    for (const tuple of data.entries()) {
        formData[tuple[0]] = tuple[1];
    }

    return formData;
}
