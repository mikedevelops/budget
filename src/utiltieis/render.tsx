import * as React from 'react';
import { Figure, ICurrencyFigure } from '../interfaces/Figure';
import LabelledFigureComponent from '../components/LabelledFigureComponent';
import AmountComponent from '../components/AmountComponent';

export function renderCurrencyWidget (
    label: string,
    entry: ICurrencyFigure
): JSX.Element {
    return (
        <div key={label} className="widget widget--currency">
            <LabelledFigureComponent label={label}>
                <AmountComponent value={entry.value} currency={entry.meta.currency}/>
            </LabelledFigureComponent>
        </div>
    );
}

export function renderPresentationWidget (
    label: string,
    entry: Figure
): JSX.Element {
    if (entry.type === 'currency') {
        return renderCurrencyWidget(label, entry);
    }
}
