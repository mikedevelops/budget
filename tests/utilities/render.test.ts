import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {renderCurrencyWidget } from '../../src/utiltieis/render';

describe('renderCurrencyWidget', () => {
    test('should render a currency widget', () => {
        const component = renderer.create(
            renderCurrencyWidget('test_currency', {
                type: 'currency',
                value: 666,
                meta: {
                    currency: 'Â£'
                }
            })
        ).toJSON();

        expect(component).toMatchSnapshot();
    });
});
