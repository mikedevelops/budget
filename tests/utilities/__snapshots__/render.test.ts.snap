// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderCurrencyWidget should render a currency widget 1`] = `
<div
  className="widget widget--currency"
>
  <div
    className="labelled-figure"
  >
    <div
      className="labelled-figure__label"
    >
      <p>
        test_currency
      </p>
    </div>
    <p
      className="amount"
    >
      Â£
      666
    </p>
  </div>
</div>
`;
